common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp

```diff
@@ -1527,65 +1527,55 @@ bool ap_wlan_hal_dwpal::read_acs_report()
 
 bool ap_wlan_hal_dwpal::read_supported_channels()
 {
-    char *reply = nullptr;
-    if (!dwpal_send_cmd("GET_HW_FEATURES", &reply)) {
-        LOG(ERROR) << __func__ << " failed";
-        return false;
-    }
+    LOG(TRACE) << __func__ << " for interface: " << get_radio_info().iface_name;
 
-    size_t replyLen = strnlen(reply, HOSTAPD_TO_DWPAL_MSG_LENGTH);
-    for (size_t i = 0; i < replyLen; i++) {
-        if (reply[i] == ')') {
-            reply[i] = ' ';
-        }
+    // Clear the supported channels vector
+    m_radio_info.supported_channels.clear();
+    // Resize the supported channels vector
+    m_radio_info.supported_channels.resize(MAX_SUPPORTED_20M_CHANNELS);
+    bwl::nl80211_client::radio_info radio_info;
+    bool ret = false;
+    if (!(ret = m_nl80211_client->get_radio_info(get_radio_info().iface_name, radio_info))) {
+        LOG(TRACE) << "Failed to get channels info";
+        return ret;
     }
 
-    LOG(DEBUG) << "GET_HW_FEATURES replylen=" << (int)replyLen;
-    LOG(DEBUG) << "GET_HW_FEATURES reply=\n" << reply;
-
-    size_t numOfValidArgs[2] = {0};
-    int chan                 = -1;
-    char dfs_state[50]       = {0};
-
-    FieldsToParse fieldsToParse[] = {
-        {(void *)&chan, &numOfValidArgs[0], DWPAL_INT_PARAM, "chan=", 0},
-        {(void *)dfs_state, &numOfValidArgs[1], DWPAL_STR_PARAM,
-         "(DFS state = ", sizeof(dfs_state)},
-        /* Must be at the end */
-        {NULL, NULL, DWPAL_NUM_OF_PARSING_TYPES, NULL, 0}};
-
-    rsize_t dmaxLen = (rsize_t)replyLen;
-    char *p2str     = nullptr;
-    char *lineMsg   = strtok_s(reply, &dmaxLen, "\n", &p2str);
-    int i           = 0;
-    while (lineMsg != NULL) {
-        size_t lineLen = strnlen(lineMsg, HOSTAPD_TO_DWPAL_MSG_LENGTH);
-
-        // TODO: fieldsToParse.field should be contained in one struct and struct size should be
-        //       passed to the dwpal_string_to_struct_parse - a security fix for multiline reply
-        if (dwpal_string_to_struct_parse(lineMsg, lineLen, fieldsToParse, sizeof(dfs_state)) ==
-            DWPAL_FAILURE) {
-            LOG(ERROR) << "DWPAL parse error ==> Abort";
-            return false;
+    int idx = 0;
```

Comment:
You should not use a vector. This function can be so much simpler if you just use the same data structure used by the nl80211 client implementation:
```
/**
 * Channels supported in this band (obtained from NL80211_BAND_ATTR_FREQS).
 * Map key is the channel number and map value is the channel information.
 */
 std::unordered_map<uint8_t, channel_info> supported_channels;
```

Reply:
As agreed currently the supported_channels will be stored as a vector
and in the son_slave_thread where ap radio basic capabilities tlv is created the channels will be stored as a map of channels and operating classes